Remove LazyInitializer usage from corelib

stephentoub opened the pull request:
It's fine for higher layers, but in corelib we don't want to forcibly prevent LazyInitializer from being trimmed nor pay the overheads of additional cached delegates.

cc: @jkotas

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:
```diff
@@ -37,20 +37,21 @@ private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOptio
+            if (createDirectory is null)
+            {
+                Type dirType = Type.GetType("System.IO.Directory, System.IO.FileSystem, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: true)!;
+                MethodInfo mi = dirType.GetTypeInfo().GetDeclaredMethod("CreateDirectory")!;
```
A bit unrelated but this will mark/keep all methods on System.IO.Directory. You could remove GetTypeInfo() to avoid marking also private stuff but perhaps the type should be moved to SPC.

stephentoub replied:
Thanks. Fixed in https://github.com/dotnet/runtime/pull/44414.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs:
```diff
@@ -78,24 +78,23 @@ private bool InitializeBinaryFormatter()
-                throwOnError: true)!);
+            Interlocked.CompareExchange(
+                ref s_binaryFormatterType,
+                Type.GetType("System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: true),
```
This will run the expensive `Type.GetType` even when `s_binaryFormatterType` is initialized already

stephentoub replied:
InitializeBinaryFormatter is only called when _binaryFormatter is null.

jkotas replied:
`_binaryFormatter` is instance field. It needs initialization per each ResourceReader instance. `s_binaryFormatterType` is global state. It needs initializing just once (it is what the original code did at least).

jkotas replied:
It would be nice to structure the initialization as:
```
if (VolatileRead(ref s_binaryFormatterType) is null || Volatile.Read(ref s_deserializeMethod) is null)
{
      // This pattern can be analyzer by IL linker. It will fix the IL linker warnings that the current code hits.
     Type type = GetType(...);
      MethodInfo method = type.GetMethod(...);

      Interlocked.CompareExchange(ref s_binaryFormatterType, ...);
      Interlocked.CompareExchange(ref s_deserializeMethod, ...);
}
```

stephentoub replied:
Ugh, you're right, will fix. Thanks.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs:
```diff
@@ -78,24 +78,23 @@ private bool InitializeBinaryFormatter()
-                throwOnError: true)!);
+            Interlocked.CompareExchange(
+                ref s_binaryFormatterType,
+                Type.GetType("System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: true),
```
Nit: The fully qualified type name can be just "System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters". The strong name, etc. is ignored by the .NET Core loader.

jkotas reviewed (approved):
LGTM

stephentoub committed:
Remove LazyInitializer usage from corelib

It's fine for higher layers, but in corelib we don't want to forcibly prevent LazyInitializer from being trimmed nor pay the overheads of additional cached delegates.
